File: apps/mobile/ios/Sources/FlagshipUI/ViewModels/WatchDelegateViewModel.swift
import Foundation
import Observation
import CryptoKit
import Flagship
import FlagshipAPI
import FlagshipCore

/// Drives the "Quick approve from Apple Watch" toggle (docs/watch-delegate-
/// key-design.md §4).
///
/// ON  → mint a fresh `.userPresence` delegate key on THIS device, have the
///       IRK attest it (one Face ID prompt), and register the IRK-signed
///       `WatchDelegateKey` with .com (scoped boot-approval, 7-day TTL).
/// OFF → IRK-sign a `RevokeWatchDelegate`, POST it, and clear the local key.
///
/// Default-OFF: nothing happens until the user flips it on. The IRK stays
/// fully biometric-gated; only this separate, boot-approval-only key is
/// minted with a laxer policy so a later Watch-driven boot approval is silent.
@Observable
@MainActor
public final class WatchDelegateViewModel {

    public enum Phase: Equatable, Sendable {
        case idle
        case loading
        case enabling
        case disabling
        case failed(String)
    }

    /// 7-day default TTL, matching the design + the cloud convention.
    public static let defaultTtlMs: Int64 = 7 * 24 * 60 * 60 * 1000

    public private(set) var phase: Phase = .idle
    /// True when an active delegate is registered for this account.
    public private(set) var isEnabled: Bool = false
    /// Expiry of the active delegate, for the "renew by <date>" hint.
    public private(set) var expiresAt: Int64?

    private let server: FlagshipServerClient
    private let username: () -> String?
    private let signer: @MainActor (String) async throws -> Curve25519.Signing.PrivateKey
    private let delegateKeyProvider: @MainActor () throws -> Curve25519.Signing.PrivateKey
    private let loadGrantId: () -> String?
    /// nil clears the local delegate entirely (key + grantId).
    private let saveGrantId: (String?) -> Void
    private let now: () -> Int64
    private let grantIdGen: () -> String

    public init(
        server: FlagshipServerClient,
        username: @escaping () -> String?,
        signer: (@MainActor (String) async throws -> Curve25519.Signing.PrivateKey)? = nil,
        delegateKeyProvider: (@MainActor () throws -> Curve25519.Signing.PrivateKey)? = nil,
        loadGrantId: (() -> String?)? = nil,
        saveGrantId: ((String?) -> Void)? = nil,
        now: (() -> Int64)? = nil,
        grantIdGen: (() -> String)? = nil
    ) {
        self.server = server
        self.username = username
        self.signer = signer ?? { reason in try await Keystore.deriveIRK(reason: reason) }
        self.delegateKeyProvider = delegateKeyProvider ?? { try Keystore.loadOrCreateWatchDelegateKey() }
        self.loadGrantId = loadGrantId ?? { Keystore.watchDelegateGrantId() }
        self.saveGrantId = saveGrantId ?? { id in
            if let id {
                try? Keystore.setWatchDelegateGrantId(id)
            } else {
                Keystore.clearWatchDelegate()
            }
        }
        self.now = now ?? { Int64(Date().timeIntervalSince1970 * 1000) }
        self.grantIdGen = grantIdGen ?? { UUID().uuidString }
    }

    /// Reconcile the toggle with the server's truth. The cloud lists only
    /// delegates that still verify under the current IRK, so a delegate
    /// orphaned by an IRK rotation reads back as "off" here.
    public func load() async {
        guard let user = username(), !user.isEmpty else { isEnabled = false; return }
        phase = .loading
        do {
            let list = try await server.listWatchDelegates(username: user)
            let active = list.delegates.first { $0.expiresAt > now() }
            isEnabled = active != nil
            expiresAt = active?.expiresAt
            phase = .idle
        } catch {
            // A read failure leaves the last-known state; surface nothing
            // destructive (the toggle just can't confirm right now).
            phase = .idle
        }
    }

    public func enable() async {
        guard let user = username(), !user.isEmpty else {
            phase = .failed("No active account on this device.")
            return
        }
        phase = .enabling
        let delegateKey: Curve25519.Signing.PrivateKey
        do {
            delegateKey = try delegateKeyProvider()
        } catch {
            phase = .failed("Couldn't create the Watch key: \(error.localizedDescription)")
            return
        }
        let irk: Curve25519.Signing.PrivateKey
        do {
            irk = try await signer("Allow your Watch to approve boots")
        } catch {
            phase = .failed("Couldn't access your account key: \(error.localizedDescription)")
            return
        }
        let issuedAt = now()
        let expiresAt = issuedAt + Self.defaultTtlMs
        let grantId = grantIdGen()
        let delegatePubHex = HexUtil.encode(delegateKey.publicKey.rawRepresentation)
        let env = WatchDelegateKeyEnvelope(
            grantId: grantId,
            username: user,
            delegatePubKeyHex: delegatePubHex,
            scopes: [WatchDelegateKeyEnvelope.bootApprovalScope],
            issuedAt: issuedAt,
            expiresAt: expiresAt
        )
        let sig: Data
        do {
            sig = try env.sign(with: irk)
        } catch {
            phase = .failed("Couldn't sign: \(error.localizedDescription)")
            return
        }
        do {
            let res = try await server.mintWatchDelegate(
                username: user,
                body: WatchDelegateMintRequest(
                    grant: .init(
                        grantId: grantId,
                        username: user,
                        delegatePubKey: delegatePubHex,
                        scopes: [WatchDelegateKeyEnvelope.bootApprovalScope],
                        issuedAt: issuedAt,
                        expiresAt: expiresAt
                    ),
                    signature: HexUtil.encode(sig)
                )
            )
            saveGrantId(res.grantId)
            isEnabled = true
            self.expiresAt = res.expiresAt
            phase = .idle
        } catch let error as ScreensClientError {
            // UX-B — plain language, no raw status code or server message.
            phase = .failed(error.errorDescription ?? "That didn't work. Try again in a moment.")
        } catch {
            phase = .failed("Couldn't reach the server. Check your connection and try again.")
        }
    }

    public func disable() async {
        guard let user = username(), !user.isEmpty else {
            phase = .failed("No active account on this device.")
            return
        }
        // Without a known grantId we can't sign a targeted revoke. Treat the
        // toggle as already-off and clear any local key as a belt-and-braces.
        guard let grantId = loadGrantId() else {
            saveGrantId(nil)
            isEnabled = false
            expiresAt = nil
            phase = .idle
            return
        }
        phase = .disabling
        let irk: Curve25519.Signing.PrivateKey
        do {
            irk = try await signer("Stop allowing your Watch to approve boots")
        } catch {
            phase = .failed("Couldn't access your account key: \(error.localizedDescription)")
            return
        }
        let issuedAt = now()
        let env = RevokeWatchDelegateEnvelope(grantId: grantId, username: user, issuedAt: issuedAt)
        let sig: Data
        do {
            sig = try env.sign(with: irk)
        } catch {
            phase = .failed("Couldn't sign: \(error.localizedDescription)")
            return
        }
        do {
            try await server.revokeWatchDelegate(
                username: user,
                body: WatchDelegateRevokeRequest(
                    request: .init(grantId: grantId, username: user, issuedAt: issuedAt),
                    signature: HexUtil.encode(sig)
                )
            )
        } catch {
            // Even if the server call fails, clear the local key so this
            // device stops being able to sign — the server's list re-verify +
            // TTL are the backstops. Report nothing destructive.
        }
        saveGrantId(nil)
        isEnabled = false
        expiresAt = nil
        phase = .idle
    }
}

File: apps/mobile/android/app/src/main/java/com/flagshipserver/app/viewmodels/WatchDelegateViewModel.kt
// Kotlin mirror of FlagshipUI's WatchDelegateViewModel. Drives the "Quick
// approve from watch" toggle (docs/watch-delegate-key-design.md §4).
//
// ON  → mint a device delegate key (no biometric gate), IRK-attest it (one
//       biometric prompt), register the WatchDelegateKey with .com.
// OFF → IRK-sign a RevokeWatchDelegate, POST it, clear the local key.
//
// Default-OFF. The IRK stays fully biometric-gated; only this separate,
// boot-approval-only key is minted without the biometric gate so a later
// watch-driven boot approval is silent.

package com.flagshipserver.app.viewmodels

import androidx.lifecycle.ViewModel
import com.flagshipserver.app.api.FlagshipServerClient
import com.flagshipserver.app.api.WatchDelegateMintRequest
import com.flagshipserver.app.api.WatchDelegateRevokeRequest
import com.flagshipserver.app.core.HexUtil
import com.flagshipserver.app.core.HttpException
import com.flagshipserver.app.core.RevokeWatchDelegate
import com.flagshipserver.app.core.WatchDelegateKey
import com.flagshipserver.app.keystore.Keystore
import com.google.crypto.tink.subtle.Ed25519Sign
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import java.util.UUID

sealed interface WatchDelegatePhase {
    data object Idle : WatchDelegatePhase
    data object Loading : WatchDelegatePhase
    data object Enabling : WatchDelegatePhase
    data object Disabling : WatchDelegatePhase
    data class Failed(val message: String) : WatchDelegatePhase
}

class WatchDelegateViewModel(
    private val server: FlagshipServerClient,
    private val username: () -> String?,
    /** Pluggable for tests; default uses the Keystore-backed IRK (biometric). */
    // Slice D — minting/revoking a watch delegate grant is SENSITIVE
    // (watchDelegates.ts gates it on the admin master root): sign with the admin
    // root when held, else the owner IRK (legacy). Canonical bytes unchanged.
    private val signer: suspend (String) -> Ed25519Sign = { r -> Keystore.adminSigningKey(r) },
    /** Ensures the device delegate key exists + returns its pubkey hex. */
    private val provisionDelegatePubHex: () -> String = {
        Keystore.loadOrCreateWatchDelegateKey()
        Keystore.watchDelegatePubHex() ?: error("delegate key missing after create")
    },
    private val loadGrantId: () -> String? = { Keystore.watchDelegateGrantId() },
    /** null clears the local delegate entirely (key + grantId). */
    private val saveGrantId: (String?) -> Unit = { id ->
        if (id != null) Keystore.setWatchDelegateGrantId(id) else Keystore.clearWatchDelegate()
    },
    private val now: () -> Long = { System.currentTimeMillis() },
    private val grantIdGen: () -> String = { UUID.randomUUID().toString() },
) : ViewModel() {

    private val _phase = MutableStateFlow<WatchDelegatePhase>(WatchDelegatePhase.Idle)
    val phase: StateFlow<WatchDelegatePhase> = _phase.asStateFlow()

    private val _isEnabled = MutableStateFlow(false)
    val isEnabled: StateFlow<Boolean> = _isEnabled.asStateFlow()

    private val _expiresAt = MutableStateFlow<Long?>(null)
    val expiresAt: StateFlow<Long?> = _expiresAt.asStateFlow()

    companion object {
        /** 7-day default TTL, matching the design + the cloud convention. */
        const val DEFAULT_TTL_MS = 7L * 24 * 60 * 60 * 1000
    }

    /** Reconcile the toggle with the server's truth. The cloud lists only
     *  delegates that still verify under the current IRK, so a delegate
     *  orphaned by an IRK rotation reads back as off. */
    suspend fun load() {
        val user = username()
        if (user.isNullOrEmpty()) { _isEnabled.value = false; return }
        _phase.value = WatchDelegatePhase.Loading
        try {
            val list = server.listWatchDelegates(user)
            val active = list.delegates.firstOrNull { it.expiresAt > now() }
            _isEnabled.value = active != null
            _expiresAt.value = active?.expiresAt
        } catch (_: Throwable) {
            // A read failure leaves the last-known state.
        }
        _phase.value = WatchDelegatePhase.Idle
    }

    suspend fun enable() {
        val user = username()
        if (user.isNullOrEmpty()) {
            _phase.value = WatchDelegatePhase.Failed("No active account on this device.")
            return
        }
        _phase.value = WatchDelegatePhase.Enabling
        val pubHex: String
        try {
            pubHex = provisionDelegatePubHex()
        } catch (e: Throwable) {
            _phase.value = WatchDelegatePhase.Failed("Couldn't create the watch key: ${e.message}")
            return
        }
        val irk: Ed25519Sign
        try {
            irk = signer("Allow your watch to approve boots")
        } catch (e: Throwable) {
            _phase.value = WatchDelegatePhase.Failed("Couldn't access your account key: ${e.message}")
            return
        }
        val issuedAt = now()
        val expiresAt = issuedAt + DEFAULT_TTL_MS
        val grantId = grantIdGen()
        val scopes = listOf(WatchDelegateKey.BOOT_APPROVAL_SCOPE)
        val sig = WatchDelegateKey.sign(irk, grantId, user, pubHex, scopes, issuedAt, expiresAt)
        try {
            val res = server.mintWatchDelegate(
                user,
                WatchDelegateMintRequest(
                    grant = WatchDelegateMintRequest.Grant(
                        grantId = grantId,
                        username = user,
                        delegatePubKey = pubHex,
                        scopes = scopes,
                        issuedAt = issuedAt,
                        expiresAt = expiresAt,
                    ),
                    signature = HexUtil.encode(sig),
                ),
            )
            saveGrantId(res.grantId)
            _isEnabled.value = true
            _expiresAt.value = res.expiresAt
            _phase.value = WatchDelegatePhase.Idle
        } catch (e: HttpException) {
            _phase.value = WatchDelegatePhase.Failed("Server rejected the request (${e.status}): ${e.body}")
        } catch (e: Throwable) {
            _phase.value = WatchDelegatePhase.Failed("Couldn't reach the server: ${e.message}")
        }
    }

    suspend fun disable() {
        val user = username()
        if (user.isNullOrEmpty()) {
            _phase.value = WatchDelegatePhase.Failed("No active account on this device.")
            return
        }
        val grantId = loadGrantId()
        if (grantId == null) {
            // Nothing to target — treat as already-off + clear any local key.
            saveGrantId(null)
            _isEnabled.value = false
            _expiresAt.value = null
            _phase.value = WatchDelegatePhase.Idle
            return
        }
        _phase.value = WatchDelegatePhase.Disabling
        val irk: Ed25519Sign
        try {
            irk = signer("Stop allowing your watch to approve boots")
        } catch (e: Throwable) {
            _phase.value = WatchDelegatePhase.Failed("Couldn't access your account key: ${e.message}")
            return
        }
        val issuedAt = now()
        val sig = RevokeWatchDelegate.sign(irk, grantId, user, issuedAt)
        try {
            server.revokeWatchDelegate(
                user,
                WatchDelegateRevokeRequest(
                    request = WatchDelegateRevokeRequest.Inner(grantId, user, issuedAt),
                    signature = HexUtil.encode(sig),
                ),
            )
        } catch (_: Throwable) {
            // Even on failure, clear the local key — the server's list
            // re-verify + TTL are the backstops.
        }
        saveGrantId(null)
        _isEnabled.value = false
        _expiresAt.value = null
        _phase.value = WatchDelegatePhase.Idle
    }
}
